### PR TITLE
feat: support error early bail

### PR DESCRIPTION
**Pull request by h-a-n-a**
## Summary

This PR introduces `option.bail`.

With `option.bail` enabled, every module factorization errors will be treated as non-recoverable errors. Webpack uses `option.bail = false` as its default value. closes https://github.com/web-infra-dev/rspack/issues/5082

This PR also refactors current empty dependency implementation:
image

Up coming PR:
Rspack converts resolve errors directly into recoverable errors, this should be aligned with webpack (i.e using `option.bail` to determine whether it is a recoverable error.)

## Test Plan

Enabled `webpack-test/cases/parsing/issue-2006/test.filter.js`
Enabled more compiler test

## Require Documentation?

- [ ] No
- [X] Yes, the corresponding rspack-website PR is https://github.com/web-infra-dev/rspack-website/pull/577

### CODE REVIEW

**Comment by h-a-n-a** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #5084** <a href="https://app.graphite.dev/github/pr/web-infra-dev/rspack/5084?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/web-infra-dev/rspack/5084?utm_source=stack-comment).

File: crates/rspack_core/src/tree_shaking/optimizer.rs [hunk 1]
```diff
@@ -493,12 +493,13 @@ impl<'a> CodeSizeOptimizer<'a> {
...
-                Some(NormalModuleSource::BuiltFailed(_)) => {
-                  // We know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
-                  continue;
+                Some(NormalModuleSource::Unbuild) => {
+                  panic!("Failed to build module {module_identifier}");
                 }
                 Some(_) => {
-                  panic!("Failed to ast of {dependency_id:?}")
+                  // If module is failed to build, we know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
+                  // Also, if the referenced module of the dependency is not found, then it might failed to factorize in the first place. So let's skip it.
```
Comment (h-a-n-a) [inline crates/rspack_core/src/tree_shaking/optimizer.rs:501]: `Module` might be not available when `dependency` is available if module factorization is failed. For example, empty dependencies are not able to create corresponding modules.

File: crates/rspack_core/src/tree_shaking/optimizer.rs [hunk 2]
```diff
@@ -641,11 +642,13 @@ impl<'a> CodeSizeOptimizer<'a> {
...
-        if matches!(ast_or_source, NormalModuleSource::BuiltFailed(_)) {
-          // We know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
-          continue;
+
+        if matches!(ast_or_source, NormalModuleSource::Unbuild) {
+          panic!("Failed to build module {cur}");
         } else {
-          panic!("Failed to resolve {dep:?}")
+          // If module is failed to build, we know that the build output can't run, so it is alright to generate a wrong tree-shaking result.
+          // Also, if the referenced module of the dependency is not found, then it might failed to factorize in the first place. So let's skip it.
```
Comment (h-a-n-a) [inline crates/rspack_core/src/tree_shaking/optimizer.rs:650]: Same as above

File: crates/rspack_core/src/module_graph/mod.rs [hunk 3]
```diff
@@ -20,7 +20,7 @@ use crate::{
 };
 
 // TODO Here request can be used JsWord
-pub type ImportVarMap = HashMap<String /* request */, String /* import_var */>;
+pub type ImportVarMap = HashMap<Option<String> /* request */, String /* import_var */>;
```
Comment (h-a-n-a) [inline crates/rspack_core/src/module_graph/mod.rs:23]: Create the `import_var` even if module does not exist.

**Review by ahabhgk — APPROVED** [review]

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**